### DISCUSSION TITLE
ESP32P4

### DISCUSSION THREAD

**Discussion by tobozo** [lovyan03/LovyanGFX#597] [Device Configurations]
For anyone feeling adventurous, there's a special branch with basic SPI/I2C support for ESP32P4:

~~https://github.com/lovyan03/LovyanGFX/tree/esp32p4~~

edit: now merged to the [develop](https://github.com/lovyan03/LovyanGFX/tree/develop) branch

**Comment by troyhacks** [comment]
I recently used LovyanGFX on my ESP32-P4 project porting WLED-MM to WLED-MM-P4.

![PXL_20251106_192217465.MP.jpg](https://github.com/user-attachments/assets/816090be-d6d3-4c1a-80bb-08d60c3f4b37)

More details and some videos here:

https://www.reddit.com/r/WLED/s/MKY6DCTdcs

This basically uses Sprites with my own defined P4 screens to draw with the Pixel Processing Accelerator.

I make RGB888 framebuffers in LovyanGFX and then the PPA "Scale Rotate Mirror" operation to composite them into my layout.

The only modification I made was to tell it to add MALLOC_CAPS_DMA and MALLOC_CAPS_CACHE_ALIGNED for the PSRAM allocator to make it go a bit faster on the P4 as you can have DMA PSRAM on the P4. Not strictly needed but worked better with these options.

Otherwise no changes to the code, it just worked. I have the Info Bar and the Button Bar as separate sprites and paint them as required.

**Reply by tobozo** [reply]
neat :heart_eyes: 

can you share your LGFX configuration for that device?

btw there's an experimental PPA layer for lgfx: https://github.com/tobozo/LGFX_PPA it comes with a PPASprite object inheriting from LGFX_Sprite and implementing the memory alignment required by ppa operations.

**Reply by troyhacks** [reply]
I didn't really "do" anything, it just worked.

I made a sprite with RGB888 and then used the PPA's SRM to composite it with "get buffer" or whatever for the sprite.

**Reply by tobozo** [reply]
> I didn't really "do" anything, it just worked.

oh, so this means the PPA -> Bus communication is handled either by Arduino_GFX or waveshare/esp_lcd_st7703 idf component?

**Comment by troyhacks** [comment]
Yes. I'm using esp_lcd_dpi_panel to control my displays.

The general code pattern: 

```
static LGFX_Sprite myFramebuffer;
...
myFramebuffer.setPsram(true);
myFramebuffer.setColorDepth(24);
myFramebuffer.createSprite(WLEDMM_DISPLAY_W, 100);
...
<set up the PPA SRM handle, etc>
<draw to myFramebuffer with library commands>
...
ppa_srm_oper_config_t srm_config = {};
srm_config.in.srm_cm = PPA_SRM_COLOR_MODE_RGB888;
srm_config.out.srm_cm = PPA_SRM_COLOR_MODE_RGB888;
srm_config.in.buffer = (uint8_t*)myFramebuffer.getBuffer();
srm_config.in.pic_w = myFramebuffer.width();
srm_config.in.pic_h = myFramebuffer.height();
srm_config.in.block_w = srm_config.in.pic_w;
srm_config.in.block_h = srm_config.in.pic_h;
srm_config.out.block_offset_x = 0;
srm_config.out.block_offset_y = 0;
< other SRM options as required >
void* fb0_ptr = NULL;
ESP_ERROR_CHECK(esp_lcd_dpi_panel_get_frame_buffer(panel_handle, 1, &fb0_ptr));
uint8_t* fb0 = (uint8_t*)fb0_ptr;
srm_config.out.buffer = fb0;
ESP_ERROR_CHECK_WITHOUT_ABORT(ppa_do_scale_rotate_mirror(ppa_srm_handle, &srm_config));
```

And the modification for library for the P4 to be faster was to add these two MALLOC_CAP options for the P4: 

`MALLOC_CAP_DMA | MALLOC_CAP_CACHE_ALIGNED`

like:

`static inline void* heap_alloc_psram(size_t length) { return heap_caps_malloc((length + 3) & ~3, MALLOC_CAP_SPIRAM | MALLOC_CAP_8BIT | MALLOC_CAP_DMA | MALLOC_CAP_CACHE_ALIGNED); }`

And really that's all I needed!

**Reply by troyhacks** [reply]
The other thing I like using is `heap_caps_malloc_prefer()` where you can specify multiple sets of MALLOC_CAP_X option sets and they're tried in order.

This allows you to "ask for PSRAM" but if it's not present, it'll fall-thru to the next set of MALLOC_CAP options. 

Or you can ask for internal and fall back to PSRAM etc
